spark-submission-worker/src/main/java/org/apache/spark/k8s/operator/SparkAppSubmissionWorker.java

```diff
@@ -0,0 +1,147 @@
+    if (MapUtils.isNotEmpty(applicationSpec.getSparkConf())) {
+      for (String confKey : applicationSpec.getSparkConf().keySet()) {
+        effectiveSparkConf.set(confKey, applicationSpec.getSparkConf().get(confKey));
+      }
+    }
+    if (MapUtils.isNotEmpty(confOverrides)) {
+      for (Map.Entry<String, String> entry : confOverrides.entrySet()) {
+        effectiveSparkConf.set(entry.getKey(), entry.getValue());
+      }
+    }
+    effectiveSparkConf.set("spark.kubernetes.namespace", app.getMetadata().getNamespace());
+    MainAppResource primaryResource = new JavaMainAppResource(Option.empty());
+    if (StringUtils.isNotEmpty(applicationSpec.getJars())) {
+      primaryResource = new JavaMainAppResource(Option.apply(applicationSpec.getJars()));
+      effectiveSparkConf.setIfMissing("spark.jars", applicationSpec.getJars());
+    } else if (StringUtils.isNotEmpty(applicationSpec.getPyFiles())) {
+      primaryResource = new PythonMainAppResource(applicationSpec.getPyFiles());
+      effectiveSparkConf.setIfMissing("spark.submit.pyFiles", applicationSpec.getPyFiles());
+    } else if (StringUtils.isNotEmpty(applicationSpec.getSparkRFiles())) {
+      primaryResource = new RMainAppResource(applicationSpec.getSparkRFiles());
+    }
+    effectiveSparkConf.setIfMissing(
+        "spark.master", "k8s://https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT");
+    String appId = generateSparkAppId(app);
+    effectiveSparkConf.setIfMissing("spark.app.id", appId);
+    return SparkAppDriverConf.create(
+        effectiveSparkConf,
+        appId,
+        primaryResource,
+        applicationSpec.getMainClass(),
+        applicationSpec.getDriverArgs().toArray(new String[0]),
+        Option.apply(applicationSpec.getProxyUser()));
+  }
+
+  protected SparkAppResourceSpec buildResourceSpec(
+      SparkAppDriverConf kubernetesDriverConf, KubernetesClient client) {
+    KubernetesDriverBuilder builder = new KubernetesDriverBuilder();
+    KubernetesDriverSpec kubernetesDriverSpec =
+        builder.buildFromFeatures(kubernetesDriverConf, client);
+    return new SparkAppResourceSpec(kubernetesDriverConf, kubernetesDriverSpec);
+  }
+
+  /**
+   * Generate a Spark application id. Similar to `KubernetesConf.getKubernetesAppId()`. This is
+   * deterministic per attempt per Spark App in order to ensure operator reconciliation idempotency
+   */
+  public static String generateSparkAppId(final SparkApplication app) {
+    long attemptId = ModelUtils.getAttemptId(app);
+    String preferredId = String.format("%s-%d", app.getMetadata().getName(), attemptId);
+    if (preferredId.length() > DEFAULT_ID_LENGTH_LIMIT) {
+      int preferredIdPrefixLength =
+          DEFAULT_ID_LENGTH_LIMIT - DEFAULT_HASH_BASED_IDENTIFIER_LENGTH_LIMIT - 1;
+      String preferredIdPrefix = preferredId.substring(0, preferredIdPrefixLength);
+      return generateHashBasedId(
+          preferredIdPrefix,
+          app.getMetadata().getNamespace(),
```

Comment:
This line seems to have some assumption on K8s namespace length.
- Can we support the maximum K8s namespace length here? 
- Please make a test case for `generateSparkAppId` for that maximum K8s namespace length.

Reply:
The namespace name / app name are used as values to generate hash which has a final length limit applied. Added Javadoc and one more test case to clarify this behavior.